lib/SIL/Utils/DynamicCasts.cpp

```diff
@@ -1355,6 +1355,29 @@ bool swift::canSILUseScalarCheckedCastInstructions(SILModule &M,
                                                   targetFormalType);
 }
 
+bool swift::canOptimizeToScalarCheckedCastInstructions(
+    SILFunction *func, CanType sourceType, CanType targetType,
+    CastConsumptionKind consumption) {
+  if (!canSILUseScalarCheckedCastInstructions(func->getModule(), sourceType,
+                                              targetType)) {
+    return false;
+  }
+
+  // Check if we preserve ownership for @guaranteed conversions.
```

Comment:
This comment + logic is very confusing. It took me a while to understand.
What you actually want to say is: if it's a copy-on-success cast we require that it preserves ownership, because the scalar cast copies the result instead of the argument.

IMO, it's much clearer if you re-factor the code into something like
```
  if (consumption == CastConsumptionKind::CopyOnSuccess) {
    // If it's a copy-on-success cast we require that it preserves ownership, because the scalar cast copies the result instead of the argument.
    return doesCastPreserveOwnershipForTypes(func->getModule(), sourceType, targetType);
  }
```

Reply:
I have updated the PR including both your suggestions. Thanks